actions/upgrade-containerd

```diff
@@ -18,38 +18,46 @@ from charmhelpers.fetch.ubuntu_apt_pkg import PkgVersion
 
 from charmhelpers.core.host import service_restart
 
-from charms.reactive import remove_state
+from charms.reactive import is_state, remove_state
 
-from reactive.containerd import CONTAINERD_PACKAGE
+from reactive.containerd import CONTAINERD_PACKAGE, apt_packages, configure_nvidia
 
 
 def _dry_run():
     """Determine if a new package is available."""
     apt_update(fatal=True)
-    try:
-        pkg = apt_cache()[CONTAINERD_PACKAGE]
-    except KeyError:
-        action_fail(f"{CONTAINERD_PACKAGE} not installed")
+    all_packages = [CONTAINERD_PACKAGE]
+
+    if is_state("containerd.nvidia.ready"):
+        all_packages += config("nvidia_apt_packages").split()
+
+    search = apt_packages(all_packages)
+    if CONTAINERD_PACKAGE not in search:
+        action_fail(f"{CONTAINERD_PACKAGE} not found in apt.")
         return
 
-    available, installed = map(PkgVersion, (pkg.get("version"), pkg.current_ver.get("ver_str")))
+    result = {}
+    for name, pkg in search.items():
+        available, installed = map(PkgVersion, (pkg.version, pkg.current_ver.ver_str))
+        result[f"{name}.available"] = available.version
+        result[f"{name}.installed"] = installed.version
+        result[f"{name}.upgrade-available"] = available > installed
 
-    result = {
-        "available": available.version,
-        "installed": installed.version,
-        "can-upgrade": available > installed
-    }
     action_set(result)
-    
+
+
 
 def _upgrade():
     """Do actual upgrade."""
     try:
         apt_update(fatal=True)
         apt_unhold(CONTAINERD_PACKAGE)
         apt_install(CONTAINERD_PACKAGE, fatal=True)
-
         apt_hold(CONTAINERD_PACKAGE)
+
+        if is_state("containerd.nvidia.ready"):
+            configure_nvidia(reconfigure=False)
```

Comment:
This seems a little heavy handed. If there aren't upgrades available for nvidia, I believe this would still cause a full removal and reinstall of the packages.

Reply:
hmm... too true

Reply:
https://github.com/charmed-kubernetes/charm-containerd/pull/73/commits/a08173f0dd223975b7a25143402c4564c3041fec address this specifically